### ISSUE TITLE
Message: can't modify frozen String

### ISSUE THREAD

**Issue by andy-thomas-83** [neillturner/kitchen-softlayer#1]
Hey,

I am trying to use your module but am running into some issues. I have pasted the kitchen file below as it wouldn't let me upload it. 

When I try and execute the test I get:

kitchen test DeployKeys
-----> Starting Kitchen (v1.4.2)
-----> Cleaning up any prior instances of <DeployKeys-Centos>
-----> Destroying <DeployKeys-Centos>...
       Finished destroying <DeployKeys-Centos> (0m0.00s).
-----> Testing <DeployKeys-Centos>
-----> Creating <DeployKeys-Centos>...

> > > > > > ------Exception-------
> > > > > > Class: Kitchen::ActionFailed
> > > > > > ## Message: Failed to complete #create action: [can't modify frozen String]
> > > > > > 
> > > > > > Please see .kitchen/logs/kitchen.log for more details
> > > > > > Also try running `kitchen diagnose --all` for configuration

I have also pasted the log file at the end.  Any help greatly appreciated.

Kitchen File:

---

---

driver:
    name: softlayer
    key_name: 'Andrews-key'
    ssh_key: '~/.ssh/id_rsa.pub'
    username: root
    datacenter: ams01
    private_server_only: true
    server_name: 'myserver-test'
    hostname: 'kitchen-test'
    domain: ibm.local
    flavor_id: m1.tiny
     #  image_id: '3b235124-a190-40b5-9720-c020e61b99e1'
    private_network_only: true
    cpu: 1,
    ram: 1024,

provisioner:
  name: ansible_playbook
  roles_path: roles

platforms:
- name: Centos
  os_code: 'CENTOS_7_64'

suites:
- name: DeployKeys
  provisioner:
    playbook: test.yml

---

Log file:
I, [2015-10-09T21:56:06.777990 #2589]  INFO -- DeployKeys-Centos: -----> Cleaning up any prior instances of <DeployKeys-Centos>
I, [2015-10-09T21:56:06.778208 #2589]  INFO -- DeployKeys-Centos: -----> Destroying <DeployKeys-Centos>...
I, [2015-10-09T21:56:06.779804 #2589]  INFO -- DeployKeys-Centos: Finished destroying <DeployKeys-Centos> (0m0.00s).
I, [2015-10-09T21:56:06.780002 #2589]  INFO -- DeployKeys-Centos: -----> Testing <DeployKeys-Centos>
I, [2015-10-09T21:56:06.780093 #2589]  INFO -- DeployKeys-Centos: -----> Creating <DeployKeys-Centos>...
E, [2015-10-09T21:56:06.892079 #2589] ERROR -- DeployKeys-Centos: Create failed on instance <DeployKeys-Centos>.
E, [2015-10-09T21:56:06.892295 #2589] ERROR -- DeployKeys-Centos: ------Exception-------
E, [2015-10-09T21:56:06.892333 #2589] ERROR -- DeployKeys-Centos: Class: RuntimeError
E, [2015-10-09T21:56:06.892359 #2589] ERROR -- DeployKeys-Centos: Message: can't modify frozen String
E, [2015-10-09T21:56:06.892382 #2589] ERROR -- DeployKeys-Centos: ------Backtrace-------
E, [2015-10-09T21:56:06.892406 #2589] ERROR -- DeployKeys-Centos: /home/andrew/.gem/ruby/gems/fog-softlayer-0.4.7/lib/fog/softlayer/slapi.rb:69:in `gsub!'
E, [2015-10-09T21:56:06.892429 #2589] ERROR -- DeployKeys-Centos: /home/andrew/.gem/ruby/gems/fog-softlayer-0.4.7/lib/fog/softlayer/slapi.rb:69:in`sanitize_username'
E, [2015-10-09T21:56:06.892453 #2589] ERROR -- DeployKeys-Centos: /home/andrew/.gem/ruby/gems/fog-softlayer-0.4.7/lib/fog/softlayer/slapi.rb:65:in `credentialize_url'
E, [2015-10-09T21:56:06.892476 #2589] ERROR -- DeployKeys-Centos: /home/andrew/.gem/ruby/gems/fog-softlayer-0.4.7/lib/fog/softlayer/slapi.rb:33:in`slapi_request'
E, [2015-10-09T21:56:06.892500 #2589] ERROR -- DeployKeys-Centos: /home/andrew/.gem/ruby/gems/fog-softlayer-0.4.7/lib/fog/softlayer/compute.rb:136:in `request'
E, [2015-10-09T21:56:06.892523 #2589] ERROR -- DeployKeys-Centos: /home/andrew/.gem/ruby/gems/fog-softlayer-0.4.7/lib/fog/softlayer/requests/compute/get_vms.rb:23:in`get_vms'
E, [2015-10-09T21:56:06.892546 #2589] ERROR -- DeployKeys-Centos: /home/andrew/.gem/ruby/gems/fog-softlayer-0.4.7/lib/fog/softlayer/compute.rb:140:in `list_servers'
E, [2015-10-09T21:56:06.892568 #2589] ERROR -- DeployKeys-Centos: /home/andrew/.gem/ruby/gems/fog-softlayer-0.4.7/lib/fog/softlayer/models/compute/servers.rb:19:in`all'
E, [2015-10-09T21:56:06.892592 #2589] ERROR -- DeployKeys-Centos: /home/andrew/.gem/ruby/gems/kitchen-softlayer-0.0.2/lib/kitchen/driver/softlayer.rb:147:in `find_server'
E, [2015-10-09T21:56:06.892660 #2589] ERROR -- DeployKeys-Centos: /home/andrew/.gem/ruby/gems/kitchen-softlayer-0.0.2/lib/kitchen/driver/softlayer.rb:77:in`create'
E, [2015-10-09T21:56:06.892696 #2589] ERROR -- DeployKeys-Centos: /usr/local/share/gems/gems/test-kitchen-1.4.2/lib/kitchen/instance.rb:424:in `public_send'
E, [2015-10-09T21:56:06.892720 #2589] ERROR -- DeployKeys-Centos: /usr/local/share/gems/gems/test-kitchen-1.4.2/lib/kitchen/instance.rb:424:in`block in perform_action'
E, [2015-10-09T21:56:06.892742 #2589] ERROR -- DeployKeys-Centos: /usr/local/share/gems/gems/test-kitchen-1.4.2/lib/kitchen/instance.rb:488:in `call'
E, [2015-10-09T21:56:06.892765 #2589] ERROR -- DeployKeys-Centos: /usr/local/share/gems/gems/test-kitchen-1.4.2/lib/kitchen/instance.rb:488:in`synchronize_or_call'
E, [2015-10-09T21:56:06.892788 #2589] ERROR -- DeployKeys-Centos: /usr/local/share/gems/gems/test-kitchen-1.4.2/lib/kitchen/instance.rb:453:in `block in action'
E, [2015-10-09T21:56:06.892811 #2589] ERROR -- DeployKeys-Centos: /usr/share/ruby/benchmark.rb:281:in`measure'
E, [2015-10-09T21:56:06.892834 #2589] ERROR -- DeployKeys-Centos: /usr/local/share/gems/gems/test-kitchen-1.4.2/lib/kitchen/instance.rb:452:in `action'
E, [2015-10-09T21:56:06.892857 #2589] ERROR -- DeployKeys-Centos: /usr/local/share/gems/gems/test-kitchen-1.4.2/lib/kitchen/instance.rb:424:in`perform_action'
E, [2015-10-09T21:56:06.892881 #2589] ERROR -- DeployKeys-Centos: /usr/local/share/gems/gems/test-kitchen-1.4.2/lib/kitchen/instance.rb:352:in `create_action'
E, [2015-10-09T21:56:06.892904 #2589] ERROR -- DeployKeys-Centos: /usr/local/share/gems/gems/test-kitchen-1.4.2/lib/kitchen/instance.rb:341:in`block in transition_to'
E, [2015-10-09T21:56:06.892942 #2589] ERROR -- DeployKeys-Centos: /usr/local/share/gems/gems/test-kitchen-1.4.2/lib/kitchen/instance.rb:340:in `each'
E, [2015-10-09T21:56:06.892967 #2589] ERROR -- DeployKeys-Centos: /usr/local/share/gems/gems/test-kitchen-1.4.2/lib/kitchen/instance.rb:340:in`transition_to'
E, [2015-10-09T21:56:06.892992 #2589] ERROR -- DeployKeys-Centos: /usr/local/share/gems/gems/test-kitchen-1.4.2/lib/kitchen/instance.rb:160:in `verify'
E, [2015-10-09T21:56:06.893016 #2589] ERROR -- DeployKeys-Centos: /usr/local/share/gems/gems/test-kitchen-1.4.2/lib/kitchen/instance.rb:189:in`block in test'
E, [2015-10-09T21:56:06.893040 #2589] ERROR -- DeployKeys-Centos: /usr/share/ruby/benchmark.rb:281:in `measure'
E, [2015-10-09T21:56:06.893064 #2589] ERROR -- DeployKeys-Centos: /usr/local/share/gems/gems/test-kitchen-1.4.2/lib/kitchen/instance.rb:185:in`test'
E, [2015-10-09T21:56:06.893087 #2589] ERROR -- DeployKeys-Centos: /usr/local/share/gems/gems/test-kitchen-1.4.2/lib/kitchen/command.rb:176:in `public_send'
E, [2015-10-09T21:56:06.893111 #2589] ERROR -- DeployKeys-Centos: /usr/local/share/gems/gems/test-kitchen-1.4.2/lib/kitchen/command.rb:176:in`block (2 levels) in run_action'
E, [2015-10-09T21:56:06.893159 #2589] ERROR -- DeployKeys-Centos: ----------------------

**Comment by andy-thomas-83** [comment]
I should have added. That I have set up the environment variables:

env | grep softlayer
softlayer_username=***********************
softlayer_api_key=***********************

**Comment by andy-thomas-83** [comment]
Applied the mods from :

NOTE: provision_script parameter needs pull request #73

Moved the os_code back into the main block and its working :) 

  platforms:
    name: Centos 
    os_code: 'CENTOS_7_64'

I guess it doesn't support multiple os'es declared by platforms?

Thanks for writing this - very much appreciated.

**Comment by neillturner** [comment]
My understanding is you specify under 

``` yaml
driver: 
 os_code: 'CENTOS_7_64'
```

and override in the suites: under 

``` yaml
   driver_config:
```

**Comment by neillturner** [comment]
Under platform you should be able to do 
platforms:

``` yaml
- name: Centos
  driver_plugin: softlayer
  driver_config:
    os_code: 'CENTOS_7_64'
```

**Comment by neillturner** [comment]
the can't modify frozen String error is in fog-softlayer and should be fixed if you apply pull request [#73](https://github.com/fog/fog-softlayer/pull/73)

**Comment by neillturner** [comment]
closing. believe this is fixed upgrading to latest kitchen-softlayer and fog-softlayer
